src/bls12_381/ec.rs

```diff
@@ -194,6 +260,34 @@ macro_rules! curve_impl {
                 (*self).into()
             }
 
+            fn hash(msg: &[u8]) -> Self {
```

Comment:
The return type of this should be projective, so I think this should be defined on CurveProjective instead. (We don't want to convert it into affine when the user may not need it in affine yet; let the user deal with this added cost.)